Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp

```diff
@@ -0,0 +1,818 @@
+    case Branch:
+    case MovHint:
+    case ExitOK:
+    case ZombieHint:
+    case InvalidationPoint:
+    case Check:
+    case CheckVarargs:
+    case Flush:
+    case GetLocal:
+    case SetLocal:
+    case GetButterfly:
+    case CheckArray:
+    case AssertNotEmpty:
+    case CheckStructure:
+    case FilterCallLinkStatus:
+    case ArrayifyToStructure:
+    case NewArrayWithConstantSize:
+    case NewArrayWithSize:
+    case ValueToInt32:
+    case ArithAdd:
+    case ArithSub:
+    case ArithMul:
+    case ArithDiv:
+    case ArithMod:
+    case ArithBitAnd:
+    case ArithBitOr:
+    case ArithBitNot:
+    case ArithBitRShift:
+    case ArithBitLShift:
+    case ArithBitXor:
+    case BitURShift:
+    case CompareLess:
+    case CompareEq:
+    case PutByVal:
+    case PutByValAlias:
+    case GetByVal: {
+        m_graph.doToChildrenWithCheck(node, [&](Edge& edge) {
+            if (isNodeCloneable(cloneAble, edge.node()))
+                return IterationStatus::Continue;
+            result = false;
+            return IterationStatus::Done;
+        });
+        break;
+    }
+    default:
+        result = false;
+    }
+
+    if (result)
+        cloneAble.add(node);
+    return result;
+}
+
+Node* LoopUnrollingPhase::cloneNode(UncheckedKeyHashMap<Node*, Node*>& nodeClones, BasicBlock* into, Node* node)
+{
+    ASSERT(node);
+    auto itr = nodeClones.find(node);
+    if (itr != nodeClones.end())
+        return itr->value;
+    Node* result = cloneNodeImpl(nodeClones, into, node);
+    ASSERT(result);
+    nodeClones.add(node, result);
+    return result;
+}
+
+Node* LoopUnrollingPhase::cloneNodeImpl(UncheckedKeyHashMap<Node*, Node*>& nodeClones, BasicBlock* into, Node* node)
+{
+#if ASSERT_ENABLED
+    m_graph.doToChildren(node, [&](Edge& e) {
+        ASSERT(e.node()->owner == node->owner);
+    });
+#endif
+
+    auto cloneEdge = [&](Edge& edge) ALWAYS_INLINE_LAMBDA {
+        return edge ? Edge(cloneNode(nodeClones, into, edge.node()), edge.useKind()) : Edge();
+    };
+
+    auto cloneAndAppend = [&](Node* node) ALWAYS_INLINE_LAMBDA {
+        Node clone = *node;
+        return into->appendNode(m_graph, WTFMove(clone));
+    };
+
+    switch (node->op()) {
+    case Phi: {
+        // Phi nodes should already be cloned in the step 2 of unrollLoop.
+        RELEASE_ASSERT_NOT_REACHED();
+        return nullptr;
+    }
+    case PutByVal:
+    case GetByVal:
+    case PutByValAlias: {
+        size_t firstChild = m_graph.m_varArgChildren.size();
+        size_t count = 0;
+        size_t expectedCount = node->op() == GetByVal ? 3 : 5;
+        m_graph.doToChildren(node, [&](Edge& edge) {
+            m_graph.m_varArgChildren.append(cloneEdge(edge));
+            ++count;
+        });
+        for (size_t i = count; i < expectedCount; ++i)
+            m_graph.m_varArgChildren.append(Edge());
+
+        Node* clone = cloneAndAppend(node);
+        clone->children.setFirstChild(firstChild);
+        return clone;
+    }
```

Comment:
I wonder if we can generalize this for `VarArgs` node case. For example, `CheckVarargs` sounds quite similar.